### PR TITLE
WIP: add: archlinux

### PR DESCRIPTION
**Pull request by SantiagoTorres**
# Checklist for Review

**NOTE:** This checklist is intended for the use of the Official Images
maintainers both to track the status of your PR and to help inform you and
others of where we're at. As such, please leave the "checking" of items to the
repository maintainers. If there is a point below for which you would like to
provide additional information or note completion, please do so by commenting
on the PR. Thanks! (and thanks for staying patient with us :heart:)

-	[x] associated with or contacted upstream?
        - (I'm a TU/part of the security team)

-	[x] does it fit into one of the common categories? ("service", "language stack", "base distribution")
-	[x] is it reasonably popular, or does it solve a particular use case well?
-	[x] does a [documentation](https://github.com/docker-library/docs/blob/master/README.md) PR exist? (should be reviewed and merged at roughly the same time so that we don't have an empty image page on the Hub for very long)
        - [yes](https://github.com/docker-library/docs/pull/1435)
-	[x] dockerization review for best practices and cache gotchas/improvements (ala [the official review guidelines](https://github.com/docker-library/official-images/blob/master/README.md#review-guidelines))?
        - I did one, albeit it may be better if someone else gives this a go.
-	[x] 2+ dockerization review?
-	[x] existing official images have been considered as a base? (ie, if `foobar` needs Node.js, has `FROM node:...` instead of grabbing `node` via other means been considered?)
-	[x] if `FROM scratch`, tarballs only exist in a single commit within the associated history?
-	[x] passes current tests? any simple new tests that might be appropriate to add? (https://github.com/docker-library/official-images/tree/master/test)

### CODE REVIEW

**Comment by SantiagoTorres** [comment]
Sigh, my bad. I must've not unstashed some changes (I had some git backflips to rebase my  changes over master)...

**Comment by SantiagoTorres** [comment]
Hi, it appears to me that this should build? (at least it does so on my local version of bashbrew). Can anyone give me pointers as to what may be wrong?

Thanks!

**Comment by SantiagoTorres** [comment]
ah, there it is. I was failing to ask for the ref :+1:

**Comment by tianon** [comment]
Here's what I get when attempting to build this:

```console
$ bashbrew build https://github.com/docker-library/official-images/raw/5258f585b46165ec78fa2ffc6a304d70188337a2/library/archlinux
Building bashbrew/cache:4b3fb9bd60a8307b7df4ae4881a07f43104b909dbc99ae9d3c7a63066d47fc37 (archlinux:latest)
error: failed building "archlinux" (tags "latest, 20190205")
exit status 1
docker ["build" "-t" "bashbrew/cache:4b3fb9bd60a8307b7df4ae4881a07f43104b909dbc99ae9d3c7a63066d47fc37" "-f" "Dockerfile" "--rm" "--force-rm" "-"] output:
Sending build context to Docker daemon  61.44kB
Step 1/8 : FROM scratch
 ---> 
Step 2/8 : ADD archlinux.tar /
 ---> 9ef7f07c8edd
Step 3/8 : RUN ldconfig && update-ca-trust && locale-gen
 ---> Running in 03deb856b9db
Removing intermediate container 03deb856b9db
OCI runtime create failed: container_linux.go:348: starting container process caused "exec: \"/bin/sh\": stat /bin/sh: no such file or directory": unknown

```

Looking at https://github.com/archlinux/archlinux-docker/commit/ddbf205adcb71047561506f54a73f80eff201c68, it appears you're running into https://github.com/docker-library/official-images/issues/1095 (TL;DR: it's unfortunately not currently possible for us to consume artifacts via Git LFS mostly due to https://github.com/git-lfs/git-lfs/issues/1322).

I note you're not using compression on that tarball -- I've had really good success compressing tarballs like this with `xz` traditionally, is that something you've considered?  It also looks like the current tarball is ~453 MB, which frankly seems absolutely massive -- is that accurate?  Is there anything in there which could reasonably be trimmed out?

**Comment by SantiagoTorres** [comment]
Hi Tianon, thanks for replying!

> Looking at [archlinux/archlinux-docker@ddbf205](https://github.com/archlinux/archlinux-docker/commit/ddbf205adcb71047561506f54a73f80eff201c68), it appears you're running into #1095 (TL;DR: it's unfortunately not currently possible for us to consume artifacts via Git LFS mostly due to [git-lfs/git-lfs#1322](https://github.com/git-lfs/git-lfs/issues/1322)).

Aha, this is what I was suspecting

> note you're not using compression on that tarball -- I've had really good success compressing tarballs like this with `xz` traditionally, is that something you've considered?

This is something I considered, but then I thought that the filesize limit was being considered on the zlib-flate version of the git object, which is a little over 100MB.

> It also looks like the current tarball is ~453 MB, which frankly seems absolutely massive -- is that accurate?

Yes, sadly it's a question of time and debate with other archlinux packagers to start trimming the fat on this image. It's something I intend on doing short-term in parallel to getting the image into the library. There's also an ongoing conversation around [what's the minimal base system](https://lists.archlinux.org/pipermail/arch-dev-public/2019-January/029435.html) which will factor in this.

> Is there anything in there which could reasonably be trimmed out?

I'm admittedly also bothered by the size. Let me also dig and make something happen before the next release. We have a CI system on our infrastructure producing this tarball which should run around the 5th of next month.

**Comment by tianon** [comment]
Any luck getting things trimmed down here?  Is this still something you're actively pursuing?

**Comment by SantiagoTorres** [comment]
> Any luck getting things trimmed down here? Is this still something you're actively pursuing?

Hi @tianon. Yes! I should be able to get back on this soon :)

**Comment by SantiagoTorres** [comment]
@tianon hey! sorry for all the delay it's been quite a while. I managed to tag and publish the release of today's build :)

**Comment by tianon** [comment]
Sorry, somehow I missed this!!

I've taken a look, and this is looking pretty sweet!  There are a couple issues with the tests that we should talk through, though:

---

Build test of #5459; ec8de817; `amd64` (`archlinux`):

```console

$ bashbrew build archlinux:latest
Building bashbrew/cache:5db6d910529cb4f90c78cdd8a43c1f490a5f3b7eeacbf0788bc486a2970328c4 (archlinux:latest)
Tagging archlinux:latest
Tagging archlinux:20190905

$ test/run.sh archlinux:latest
testing archlinux:latest
	'utc' [1/4]...passed
	'cve-2014--shellshock' [2/4]...shellshock_test.sh: line 5: grep: command not found
shellshock_test.sh: line 17: grep: command not found
shellshock_test.sh: line 28: grep: command not found
shellshock_test.sh: line 39: grep: command not found
shellshock_test.sh: line 50: grep: command not found
shellshock_test.sh: line 61: grep: command not found
shellshock_test.sh: line 72: grep: command not found
passed
	'no-hard-coded-passwords' [3/4]...error: empty password detected for 'root'
failed
	'override-cmd' [4/4]...passed

```


The following images failed at least one test:

- `archlinux`:
  - `archlinux:latest`

---

That `cve-2014--shellshock` failure is hilariously because the image contains `bash` (which the test detects and thus runs), but does not contain `grep`.  If that was an oversight, that's an easy fix, but otherwise we'll probably need to do something with that test.  I've been contemplating removing it, because it doesn't serve any real purpose these days, but it is a fun demonstration of the test framework (and has caught bugs with the `bash:devel` image before :sweat_smile:), so maybe we should update it to also detect `grep`?

The failure in `no-hard-coded-passwords` is much more interesting/relevant and is very likely actionable.  See CVE-2019-5021 for details (https://alpinelinux.org/posts/Docker-image-vulnerability-CVE-2019-5021.html is a good description of the problem). :+1:

**Comment by SantiagoTorres** [comment]
Thanks for reviewing @tianon !

>  If that was an oversight, that's an easy fix, but otherwise we'll probably need to do something with that test. I've been contemplating removing it, because it doesn't serve any real purpose these days, but it is a fun demonstration of the test framework (and has caught bugs with the bash:devel image before sweat_smile), so maybe we should update it to also detect grep?

We indeed do not explicitly include grep. We pretty much include [4 packages](https://github.com/archlinux/archlinux-docker/blob/master/packages) and its dependencies. I think adding a test to detect grep would be useful. Would you want me to give that a go?

> The failure in no-hard-coded-passwords is much more interesting/relevant and is very likely actionable. See CVE-2019-5021 for details (https://alpinelinux.org/posts/Docker-image-vulnerability-CVE-2019-5021.html is a good description of the problem). +1

This one is definitely worth fixing on our side. Thanks for the catch! I'll update the image.

**Comment by tianon** [comment]
I took a look, and I think this is the simplest solution for that test:

```diff
diff --git a/test/tests/cve-2014--shellshock/run.sh b/test/tests/cve-2014--shellshock/run.sh
index aa9ece7..03bdc4c 100755
--- a/test/tests/cve-2014--shellshock/run.sh
+++ b/test/tests/cve-2014--shellshock/run.sh
@@ -6,5 +6,10 @@ if ! docker run --rm --entrypoint bash "$1" -c 'true' &> /dev/null; then
 	exit 0
 fi
 
+if ! docker run --rm --entrypoint bash "$1" -c 'command -v grep' &> /dev/null; then
+	echo >&2 'skipping; image does not contain "grep"'
+	exit 0
+fi
+
 dir="$(readlink -f "$(dirname "$BASH_SOURCE")")"
 source "$dir/really-run.sh" "$@"
```

**Comment by SantiagoTorres** [comment]
Thanks for updating the tests. If I didn't screw up on my side things should be working :)

**Comment by tianon** [comment]
I didn't actually update the tests, just played with doing so -- if you include that diff as a commit in this PR, then our test-running script will DTRT. :smile: (Happy to do so for you if you'd rather :+1:)

However, it appears the test doesn't actually care too much because it passes anyways:

---

Build test of #5459; 1f28612b; `amd64` (`archlinux`):

```console

$ bashbrew build archlinux:latest
Building bashbrew/cache:d1f3d56da609de8edb6c1f3458057ef50d8ddfe3c222b524b947f6649ac6c6d6 (archlinux:latest)
Tagging archlinux:latest
Tagging archlinux:20191006

$ test/run.sh archlinux:latest
testing archlinux:latest
	'utc' [1/4]...passed
	'cve-2014--shellshock' [2/4]...shellshock_test.sh: line 5: grep: command not found
shellshock_test.sh: line 17: grep: command not found
shellshock_test.sh: line 28: grep: command not found
shellshock_test.sh: line 39: grep: command not found
shellshock_test.sh: line 50: grep: command not found
shellshock_test.sh: line 61: grep: command not found
shellshock_test.sh: line 72: grep: command not found
passed
	'no-hard-coded-passwords' [3/4]...passed
	'override-cmd' [4/4]...passed

```

**Comment by tianon** [comment]
I think this is looking really good!  The docs PR needs a couple minor tweaks (formatting), but I'm satisfied. :+1:

@yosifkit ? :smile:

**Comment by shibumi** [comment]
@tianon can you specify which minor tweaks you mean?

**Comment by tianon** [comment]
I mean the comments from https://github.com/docker-library/docs/pull/1435#pullrequestreview-299742233, specifically. :+1:

File: library/archlinux [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+
+Maintainers: Santiago Torres-Arias <santiago@archlinux.org> (@SantiagoTores),
+             Christian Rebischke <Chris.Rebischke@archlinux.org> (@shibumi),
+             Pierre Schmitz <pierre@archlinux.de> (@pierres)
+GitRepo: https://github.com/archlinux/archlinux-docker.git
+
+
+Tags: latest, 20191006
+GitCommit: 671e950b6f879a346fffa2f9fbaaba8ace07efae
+GitFetch: refs/tags/20191006
```
Comment (tianon) [inline library/archlinux]: Gah, looks like https://github.com/archlinux/archlinux-docker/commits/20191006 has a tarball being modified in the history. :disappointed:

>  - [ ] if `FROM scratch`, tarballs only exist in a single commit within the associated history?

Comment (SantiagoTorres) [reply library/archlinux]: ugh! I rebased the history there and removed that old tarball. Is this a reasonable fix or?

Comment (tianon) [reply library/archlinux]: Exactly perfect, LGTM! :tada: 

@yosifkit ? :innocent: :+1:

**Comment by yosifkit** [comment]
Initial diff: https://gist.github.com/docker-library-bot/0ce0c06ec7f90fcf1354dbc430c6d74f

My only comment is that you'll want to combine these two lines into one `RUN` to delete the keys in the layer that they are created:
```dockerfile
RUN pacman-key --init && pacman-key --populate archlinux
RUN rm -rf etc/pacman.d/gnupg/{openpgp-revocs.d/,private-keys-v1.d/,pugring.gpg~,gnupg.S.}*

# for example
RUN pacman-key --init && pacman-key --populate archlinux \
  && rm -rf etc/pacman.d/gnupg/{openpgp-revocs.d/,private-keys-v1.d/,pugring.gpg~,gnupg.S.}*
```

**Comment by yosifkit** [comment]
Build test of #5459; 1acfc434; `amd64` (`archlinux`):

```console

$ bashbrew build archlinux:latest
Building bashbrew/cache:ad6ab53eab82c1e6f9f740e6ca5c6b720f9de2933047e1dbd69f41708ef61a38 (archlinux:latest)
Tagging archlinux:latest
Tagging archlinux:20191006

$ test/run.sh archlinux:latest
testing archlinux:latest
	'utc' [1/4]...passed
	'cve-2014--shellshock' [2/4]...shellshock_test.sh: line 5: grep: command not found
shellshock_test.sh: line 17: grep: command not found
shellshock_test.sh: line 28: grep: command not found
shellshock_test.sh: line 39: grep: command not found
shellshock_test.sh: line 50: grep: command not found
shellshock_test.sh: line 61: grep: command not found
shellshock_test.sh: line 72: grep: command not found
passed
	'no-hard-coded-passwords' [3/4]...passed
	'override-cmd' [4/4]...passed

```

**Pull request merged**